### PR TITLE
Fix copyTo bug

### PR DESCRIPTION
**Pull request by ianopolous**
There was a bug which meant dirA.copyTo(dirB) would fail if dirA contained a file.

### CODE REVIEW

File: src/peergos/shared/user/fs/FileWrapper.java [hunk 1]
```diff
@@ -1099,28 +1099,40 @@ static boolean isLegalName(String name) {
...
-                    }
-                    return pointer.fileAccess.mkdir(state, committer, newFolderName, network, writableFilePointer(), getChildsEntryWriter(),
-                            requestedBaseSymmetricKey, isSystemFolder, crypto).thenApply(x -> {
-                        setModified();
-                        return x;
-                    });
-                })).thenCompose(version -> getUpdated(version, network));
+        return hasChildWithName(version, newFolderName, crypto.hasher, network).thenCompose(hasChild -> {
+            if (hasChild) {
+                CompletableFuture<Snapshot> error = new CompletableFuture<>();
```
Comment (kevodwyer) [inline src/peergos/shared/user/fs/FileWrapper.java]: change to Futures.errored()

File: src/peergos/shared/user/fs/FileWrapper.java [hunk 2]
```diff
@@ -1099,28 +1099,40 @@ static boolean isLegalName(String name) {
...
                                                 Crypto crypto) {
 
-        CompletableFuture<FileWrapper> result = new CompletableFuture<>();
+        return network.synchronizer.applyComplexUpdate(owner(), signingPair(),
+                (state, committer) -> mkdir(newFolderName, Optional.ofNullable(requestedBaseSymmetricKey),
+                        Optional.empty(), Optional.empty(), isSystemFolder, network, crypto, state, committer))
+                .thenCompose(version -> getUpdated(version, network));
+    }
+
+    public CompletableFuture<Snapshot> mkdir(String newFolderName,
```
Comment (kevodwyer) [inline src/peergos/shared/user/fs/FileWrapper.java]: can this be private?

File: src/peergos/shared/user/fs/FileWrapper.java [hunk 3]
```diff
@@ -1245,43 +1257,61 @@ public SigningPrivateKeyAndPublicHash signingPair() {
...
+            if (isDirectory()) {
+                byte[] newMapKey = crypto.random.randomBytes(32);
+                SymmetricKey newBaseR = SymmetricKey.random();
+                SymmetricKey newBaseW = SymmetricKey.random();
+                WritableAbsoluteCapability newCap = ((WritableAbsoluteCapability)target.getPointer().capability)
+                        .withMapKey(newMapKey)
+                        .withBaseKey(newBaseR)
+                        .withBaseWriteKey(newBaseW);
+                return getChildren(version, crypto.hasher, network).thenCompose(children -> {
+                    return target.mkdir(getName(), Optional.of(newBaseR), Optional.of(newBaseW), Optional.of(newMapKey),
```
Comment (kevodwyer) [inline src/peergos/shared/user/fs/FileWrapper.java]: return keyword not required, use simple closure form

File: src/peergos/shared/user/fs/FileWrapper.java (hunk 1 above)
Comment (ianopolous) [reply src/peergos/shared/user/fs/FileWrapper.java]: done

File: src/peergos/shared/user/fs/FileWrapper.java (hunk 2 above)
Comment (ianopolous) [reply src/peergos/shared/user/fs/FileWrapper.java]: done

File: src/peergos/shared/user/fs/FileWrapper.java (hunk 3 above)
Comment (ianopolous) [reply src/peergos/shared/user/fs/FileWrapper.java]: done

**Pull request merged**